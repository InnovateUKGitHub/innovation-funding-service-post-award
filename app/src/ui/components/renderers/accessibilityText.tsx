export const AccessibilityText: React.FunctionComponent = props => (
  <span className="govuk-visually-hidden">{props.children}</span>
);
