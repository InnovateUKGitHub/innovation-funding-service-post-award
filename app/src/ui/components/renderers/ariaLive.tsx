interface AriaLiveProps {
  children: React.ReactNode;
}

export function AriaLive(props: AriaLiveProps) {
  return <div aria-live="polite" {...props} />;
}
