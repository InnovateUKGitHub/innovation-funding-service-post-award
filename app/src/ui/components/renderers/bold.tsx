type SpanAttributes = React.HTMLAttributes<HTMLSpanElement>;

export function Bold(props: SpanAttributes) {
  return <span {...props} className="govuk-!-font-weight-bold" />;
}
