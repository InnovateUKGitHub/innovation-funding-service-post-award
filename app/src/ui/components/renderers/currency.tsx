import React from "react";
import {isNumber} from "../../../util/NumberHelper";

interface Props {
  value: number | null;
  fractionDigits?: number;
}

export const Currency: React.SFC<Props> = ({ value, fractionDigits = 2 }) => {
  if (!isNumber(value)) {
    return null;
  }

  const options = {
    style: "currency",
    currency: "GBP",
    minimumFractionDigits: fractionDigits,
    maximumFractionDigits: fractionDigits
  };

  const valToRender = new Intl.NumberFormat("en-GB", options).format(value);
  return <span style={{ whiteSpace: "nowrap" }}>{valToRender}</span>;
};
