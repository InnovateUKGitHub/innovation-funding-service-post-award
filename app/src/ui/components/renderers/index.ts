export * from "./ariaLive";
export * from "./currency";
export * from "./date";
export * from "./email";
export * from "./currency";
export * from "./messages";
export * from "./percentage";
export * from "./simpleString";
export * from "./json";
