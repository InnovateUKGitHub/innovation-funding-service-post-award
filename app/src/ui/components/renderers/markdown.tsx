import { marked } from "marked";

export interface IMarkdownProps {
  value: string;
  style?: React.CSSProperties;
}

/**
 * Renders a component that renders a string using markdown syntax
 */
export function Markdown({ value, ...props }: IMarkdownProps) {
  if (!value.length) return null;

  return <span {...props} className="govuk-body markdown" dangerouslySetInnerHTML={{ __html: marked.parse(value) }} />;
}
