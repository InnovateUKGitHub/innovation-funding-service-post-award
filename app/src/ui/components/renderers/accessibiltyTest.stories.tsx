import type { Meta, StoryObj } from "@storybook/react";

import { AccessibilityText } from "./accessibilityText";

const meta: Meta<typeof AccessibilityText> = {
  title: "IFSPA Renderer/Accessibility Text",
  tags: ["ifspa/renderer"],
  component: AccessibilityText,
};

export default meta;

type Story = StoryObj<typeof AccessibilityText>;

export const Primary: Story = {
  args: {
    children: "The quick brown fox jumps over the lazy dog.",
  },
};
