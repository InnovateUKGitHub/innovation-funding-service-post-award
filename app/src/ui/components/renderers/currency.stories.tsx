import type { Meta, StoryObj } from "@storybook/react";

import { Currency } from "./currency";

const meta: Meta<typeof Currency> = {
  title: "IFSPA Renderer/Currency",
  tags: ["ifspa/renderer"],
  component: Currency,
};

export default meta;

type Story = StoryObj<typeof Currency>;

export const Primary: Story = {
  args: {
    value: 124,
    fractionDigits: 2,
  },
};
export const RoundToZeroDecimalPlaces: Story = {
  args: {
    value: 124.91,
    fractionDigits: 0,
  },
};
export const WholeNumberWithCommas: Story = {
  args: {
    value: 4815162342,
    fractionDigits: 0,
  },
};
