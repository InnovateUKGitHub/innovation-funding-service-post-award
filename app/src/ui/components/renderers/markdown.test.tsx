import { render } from "@testing-library/react";

import { IMarkdownProps, Markdown } from "@ui/components/renderers/markdown";

describe("Markdown", () => {
  const setup = (props: IMarkdownProps) => render(<Markdown {...props} />);

  const stubMarkdownHeading = `
Heading Test
------------
`;

  describe("@renders", () => {
    it("should wrap content in <span>", () => {
      const { container } = setup({ value: "some-valid-string" });

      if (!container.firstChild) {
        throw Error("First child not found!");
      }

      expect(container.firstChild.nodeName).toBe("SPAN");
    });

    it("should have markdown class for style overrides", () => {
      const { container } = setup({ value: "some-valid-string" });

      if (!container.firstChild) {
        throw Error("First child not found!");
      }

      expect(container.firstChild).toHaveClass("markdown");
    });

    it("should return null when empty string is passed", () => {
      const { container } = setup({ value: "" });

      expect(container.firstChild).toBeNull();
    });

    it("should return styles", () => {
      const stubColorStyleValue = "red";

      const { container } = setup({ value: "non-empty-string", style: { color: stubColorStyleValue } });

      expect(container.firstChild).toHaveStyle(`color: ${stubColorStyleValue}`);
    });

    describe("with converted markdown parsed correctly", () => {
      test.each`
        name                               | trusted  | bannedStrings                | markdown
        ${"with string"}                   | ${true}  | ${[]}                        | ${"stub-content"}
        ${"with headings"}                 | ${true}  | ${[]}                        | ${stubMarkdownHeading}
        ${"with ordered lists"}            | ${true}  | ${[]}                        | ${"1. stub-ordered-item"}
        ${"with unordered lists"}          | ${true}  | ${[]}                        | ${"* stub-unordered-item"}
        ${"with html and secure enabled"}  | ${true}  | ${[]}                        | ${"<p>Hello world!</p>"}
        ${"with html and secure disabled"} | ${false} | ${["<script>", "</script>"]} | ${'<script>alert("Neil Little")</script>'}
      `("$name", ({ markdown, trusted, bannedStrings }) => {
        const { container } = setup({ value: markdown, trusted });

        // Use snapshot testing to compare Markdown output values
        expect(container.innerHTML).toMatchSnapshot();

        // Ensure that banned values will NEVER be in the markdown output
        for (const bannedString of bannedStrings) {
          expect(container.innerHTML).not.toContain(bannedString);
        }
      });
    });
  });
});
