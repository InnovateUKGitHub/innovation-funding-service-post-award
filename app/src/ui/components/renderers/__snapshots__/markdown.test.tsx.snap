// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown @renders with converted markdown parsed correctly with headings 1`] = `
"<span class=\\"govuk-body markdown\\"><h2 id=\\"heading-test\\">Heading Test</h2>
</span>"
`;

exports[`Markdown @renders with converted markdown parsed correctly with html and secure disabled 1`] = `"<span class=\\"govuk-body markdown\\"></span>"`;

exports[`Markdown @renders with converted markdown parsed correctly with html and secure enabled 1`] = `"<span class=\\"govuk-body markdown\\"><p>Hello world!</p></span>"`;

exports[`Markdown @renders with converted markdown parsed correctly with ordered lists 1`] = `
"<span class=\\"govuk-body markdown\\"><ol>
<li>stub-ordered-item</li>
</ol>
</span>"
`;

exports[`Markdown @renders with converted markdown parsed correctly with string 1`] = `
"<span class=\\"govuk-body markdown\\"><p>stub-content</p>
</span>"
`;

exports[`Markdown @renders with converted markdown parsed correctly with unordered lists 1`] = `
"<span class=\\"govuk-body markdown\\"><ul>
<li>stub-unordered-item</li>
</ul>
</span>"
`;
